File: eval/cases-ts/TN01/api/src/handlers.js
import { payoutResponse, balanceResponse, customerResponse, errorResponse } from "./types.js";

export function createPayout(req, res) {
  if (!req.get("Idempotency-Key")) {
    return res.status(400).json(errorResponse("missing_idempotency_key", "Idempotency-Key header is required"));
  }
  if (!req.body || !req.body.reference) {
    return res.status(400).json(errorResponse("missing_reference", "reference is required"));
  }
  return res.status(201).json(payoutResponse(req.body));
}

export function getPayout(req, res) {
  return res.status(200).json(payoutResponse(
    { amount: "150000", currency: "NGN", reference: "ref_123" }, req.params.id));
}

export function createRefund(req, res) {
  return res.status(201).json({
    id: "rf_01HZX",
    payoutId: req.body.payoutId,
    status: "pending",
    amount: req.body.amount,
  });
}

export function getBalance(_req, res) {
  return res.status(200).json(balanceResponse());
}

export function listTransactions(req, res) {
  let pageNum = Number(req.query.page);
  if (!Number.isFinite(pageNum) || pageNum < 1) pageNum = 1;
  let page = pageNum;
  let perPage = Number(req.query.perPage);
  if (!Number.isFinite(perPage) || perPage < 1) perPage = 25;
  if (perPage > 100) perPage = 100;
  return res.status(200).json({ data: [], page, perPage, total: 0 });
}

export function createCustomer(req, res) {
  if (!req.body || !req.body.email) {
    return res.status(400).json(errorResponse("missing_email", "email is required"));
  }
  return res.status(201).json(customerResponse(req.body));
}

export function verifyBvn(req, res) {
  if (!req.body || String(req.body.bvn || "").length !== 11) {
    return res.status(400).json(errorResponse("invalid_bvn", "bvn must be 11 digits"));
  }
  return res.status(200).json({ verified: true, firstName: "Ada", lastName: "Okafor", tier: 1 });
}

export function listBanks(_req, res) {
  return res.status(200).json([
    { code: "044", name: "Access Bank", slug: "access-bank" },
    { code: "058", name: "Guaranty Trust Bank", slug: "gtb" },
  ]);
}

export function testWebhook(_req, res) {
  res.set("X-Signature", "sha256=deadbeef");
  return res.status(202).json({ success: true, code: "queued", message: "test event queued" });
}

File: eval/cases-py/P04/api/app/handlers.py
"""Handlers for the fixture payments API.

Each returns the response body only, and the status is declared on the route
decorator - the idiomatic FastAPI shape. Errors raise HttpError, which carries
the status the caller would see.
"""

from .models import payout_response, balance_response, customer_response


class HttpError(Exception):
    def __init__(self, status, code, message):
        super().__init__(message)
        self.status, self.code, self.message = status, code, message


def create_payout(body, headers=None):
    headers = headers or {}
    if not headers.get("Idempotency-Key"):
        raise HttpError(400, "missing_idempotency_key", "Idempotency-Key header is required")
    if not body.get("reference"):
        raise HttpError(400, "missing_reference", "reference is required")
    return payout_response(body)


def get_payout(payout_id):
    return payout_response(
        {"amount": "150000", "currency": "NGN", "reference": "ref_123"}, payout_id)


def create_refund(body):
    return {
        "id": "rf_01HZX",
        "payoutId": body.get("payoutId"),
        "status": "pending",
        "amount": body.get("amount"),
    }


def get_balance():
    return balance_response()


def list_transactions(page=1, perPage=25):
    page = max(int(page or 1), 1)
    perPage = int(perPage or 25)
    if perPage < 1:
        perPage = 25
    if perPage > 100:
        perPage = 100
    return {"data": [], "page": page, "perPage": perPage, "total": 0}


def create_customer(body):
    if not body.get("email"):
        raise HttpError(400, "missing_email", "email is required")
    return customer_response(body)


def verify_bvn(body):
    if len(str(body.get("bvn", ""))) != 11:
        raise HttpError(400, "invalid_bvn", "bvn must be 11 digits")
    return {"verified": True, "firstName": "Ada", "lastName": "Okafor", "tier": 1}


def list_banks():
    return [
        {"code": "044", "name": "Access Bank", "slug": "access-bank"},
        {"code": "058", "name": "Guaranty Trust Bank", "slug": "gtb"},
    ]
